### DISCUSSION TITLE
Using appropriate Junicode weights at different sizes

### DISCUSSION THREAD

**Discussion by adunning** [psb1558/Junicode-font#92] [Show and tell]
I have been using Junicode's new weights as 'optical' sizes to achieve even type colour, avoiding the typical problem in digital typography of text that becomes lighter rather than darker at smaller sizes. In LaTeX, I've created the following definition, based on Adobe's Caption/Small Text/Text/Subhead/Display weights (cf. the [Arno Pro specimen book](https://www.adobe.com/content/dam/acom/en/products/type/pdfs/ArnoPro.pdf)). These numbers need to be adjusted to fit the medium and viewing conditions, but I've found this works fairly well with my printer. I would be interested in any suggestions to optimize this further or integrate any other weights:

```latex
\setmainfont{Junicode Two Beta}[
  UprightFont = {*-Regular},
  ItalicFont = {*-Italic},
  BoldFont = {*-Semibold},
  BoldItalicFont = {*-BoldItalic},
  Numbers={Lowercase,Proportional},
  UprightFeatures={
    SizeFeatures={
      {Size={-8.5},      Font=*-SemiExpandedMedium},
      {Size={8.6-10.99}, Font=*-Medium},
      {Size={11-21.59},  Font=*-Regular},
      {Size={21.6-},     Font=*-Light}
    },
  },
  ItalicFeatures={
    SizeFeatures={
      {Size={-8.5},      Font=*-SemiExpandedMediumItalic},
      {Size={8.6-10.99}, Font=*-MediumItalic},
      {Size={11-21.59},  Font=*-Italic},
      {Size={21.6-},     Font=*-LightItalic}
    },
  },
]
```

Compare a LaTeX configuration for Elstob's optical weights:

```latex
\setmainfont{ElstobD}[
  UprightFont = {*-Regular},
  ItalicFont = {*-Italic},
  BoldFont = {*-SemiBold},
  BoldItalicFont = {*-SemiBoldItalic},
  Numbers={Lowercase,Proportional},
  UprightFeatures={
    SizeFeatures={
      {Size={-8.5},      Font=*-8pt},
      {Size={8.6-10.99}, Font=*-10pt},
      {Size={11-14.5},   Font=*-14pt},
      {Size={14.6-21.59},Font=*-18pt},
      {Size={21.6-35.99},Font=*-Light},
      {Size={36-},       Font=*-ExtraLight}
    },
  },
  ItalicFeatures={
    SizeFeatures={
      {Size={-8.5},      Font=*-8ptItalic},
      {Size={8.6-10.99}, Font=*-10ptItalic},
      {Size={11-14.5},   Font=*-14ptItalic},
      {Size={14.6-21.59},Font=*-18ptItalic},
      {Size={21.6-35.99},Font=*-LightItalic},
      {Size={36-},       Font=*-ExtraLightItalic}
    },
  },
  BoldFeatures={
    SizeFeatures={
      {Size={-8.5},      Font=*-8ptSemiBold},
      {Size={8.6-10.99}, Font=*-10ptSemiBold},
      {Size={11-14.5},   Font=*-14ptSemiBold},
      {Size={14.6-},     Font=*-18ptSemiBold}
    },
  },
  BoldItalicFeatures={
    SizeFeatures={
      {Size={-8.5},      Font=*-8ptSemiBoldItalic},
      {Size={8.6-10.99}, Font=*-10ptSemiBoldItalic},
      {Size={11-14.5},   Font=*-14ptSemiBoldItalic},
      {Size={14.6-},     Font=*-18ptSemiBoldItalic}
    },
  }
]
```

**Comment by psb1558** [comment]
Thanks, Andrew! I look forward to taking these for a spin, and I'll let you know if any thoughts occur to me.

I don't suppose support for variable fonts is anywhere on the horizon?

**Comment by psb1558** [comment]
I like both sets of sizes, though I suppose it's inevitable that I find the optical sizes of Elstob more satisfying than the pseudo-optical sizes of Junicode. I've got to confess that I'm not a big fan of the expanded widths in the italic, and so I made the cutoff for using it lower: if I could I'd make the cutoff about a third of the way from SemiExpanded down to Regular. For some reason I also had trouble locating the fonts (I was using XeLaTeX) and had to  insert some space in the style names to make it work.

So here's my attempt:

```
\setmainfont{Junicode Two Beta}[
  UprightFont = {*-Regular},
  ItalicFont = {*-Italic},
  BoldFont = {*-Semibold},
  BoldItalicFont = {*-Bold Italic},
  Numbers={Lowercase,Proportional},
  UprightFeatures={
    SizeFeatures={
      {Size={-8.5},      Font={*-SemiExpanded Medium}},
      {Size={8.6-10.99}, Font=*-Medium},
      {Size={11-21.59},  Font=*-Regular},
      {Size={21.6-},     Font=*-Light}
    },
  },
  ItalicFeatures={
    SizeFeatures={
      {Size={-7.5},      Font={*-SemiExpanded Medium Italic}},
      {Size={7.6-10.99}, Font={*-Medium Italic}},
      {Size={11-21.59},  Font=*-Italic},
      {Size={21.6-},     Font={*-Light Italic}}
    },
  },
  BoldFeatures={
    SizeFeatures={
      {Size={-8.5},      Font={*-SemiExpanded Bold}},
      {Size={8.6-10.99}, Font=*-Bold},
      {Size={11-21.59},  Font=*-Semibold},
      {Size={21.6-},     Font=*-Medium}
    },
  },
  BoldItalicFeatures={
    SizeFeatures={
      {Size={-7.5},      Font={*-SemiExpanded Bold Italic}},
      {Size={7.6-10.99}, Font={*-Bold Italic}},
      {Size={11-21.59},  Font={*-Semibold Italic}},
      {Size={21.6-},     Font={*-Medium Italic}}
    },
  },
]
```

Reading that LuaTeX now supports variable fonts, I tried out the variable version:

```
\setmainfont{Junicode Two Beta VF}[
  UprightFont = {*-Regular},
  ItalicFont = {*-Italic},
  Numbers={Lowercase,Proportional},
  UprightFeatures={
    SizeFeatures={
      {Size={-8.5},      RawFeature={axis={wght=600,wdth=112}}},
      {Size={8.6-10.99}, RawFeature={axis={wght=500,wdth=106}}},
      {Size={11-21.59},  RawFeature={axis={wght=380,wdth=100}}},
      {Size={21.6-},     RawFeature={axis={wght=300,wdth=95}}}
    },
  },
  ItalicFeatures={
    SizeFeatures={
      {Size={-8.5},      RawFeature={axis={wght=600,wdth=112}}},
      {Size={8.6-10.99}, RawFeature={axis={wght=500,wdth=106}}},
      {Size={11-21.59},  RawFeature={axis={wght=400,wdth=100}}},
      {Size={21.6-},     RawFeature={axis={wght=300,wdth=95}}}
    },
  },
]
```

Unfortunately whatever renderer LuaTeX is using seems very buggy:

<img width="228" alt="image" src="https://user-images.githubusercontent.com/16215767/145937822-5f199343-cc7c-416a-96b9-749734e4c2e5.png">

This kind of effect on certain characters at certain weights and widths affects fonts other than Junicode, though not all. I haven't been able to find any discussions of it, so I don't know what might be causing it or how to work around it. It affects the roman face but apparently not the italic.

**Reply by adunning** [reply]
What a strange effect; I have not been able to reproduce it, so it may have been fixed now. Thank you for the improved configurations!

**Reply by adunning** [reply]
This bug seems to occur only with TTF variable fonts; OTF ones are fine as far as I can see (e.g. Source Serif 4 provides both). I've tried filing a bug report: <https://github.com/latex3/luaotfload/issues/209>.

**Reply by psb1558** [reply]
It looks to me as if the bug in luaotfload has been around for quite a while, and I'd guess there's not a lot of pressure to fix it because variable fonts aren't particularly widespread yet. Thanks for filing the bug report!

Glyphs doesn't generate cff2 variable fonts, so I'll try fontmake, which (apparently) does. I haven't had much luck building Junicode with fontmake yet (crashes with a weirdly unhelpful error message) but will see if I can track down the problem.

**Reply by psb1558** [reply]
It's going to take a while to get fontmake working with Junicode. I have managed to produce a cff2-flavored version of Elstob Regular and will test with that.

**Comment by psb1558** [comment]
I've put a new folder, variable-cff2, in the fonts folder. It contains JunicodeTwoBeta-VF.otf. I'll test this later with luaTeX, but for right now some notes:

- The family name is "Junicode Two Beta" when it should be "Junicode Two Beta VF." Sorry--I'll fix this later.
- fontmake's variable-cff2 builder doesn't seem to like virtual masters (I'll look into this and open an issue if necessary), so those are suppressed here, with the result that the interpolations for this version won't be as good as in the TrueType version.
- By way of testing, all I can say is that [FontGoggles](https://fontgoggles.org/) likes this font fine, but [Samsa](https://github.com/Lorp/samsa) does not. (I don't know whether Samsa can handle cff2-flavored variable fonts. More than a few tools do not.)
- This font is extremely experimental, and for now only intended for testing in luaTeX.

**Comment by psb1558** [comment]
A much better cff2-flavored variable font now in fonts/variable-cff2. I haven't tested with luaTeX yet.

**Comment by psb1558** [comment]
Tested in luaTeX with this file:

```
\documentclass{article}
\usepackage{fontspec}

\setmainfont{JunicodeTwoBeta-Regular-VF.otf}[
  Numbers={Lowercase,Proportional},
  UprightFeatures={
    SizeFeatures={
      {Size={-8.5},          RawFeature={axis={wght=600,wdth=112}}},
      {Size={8.6-10.99}, RawFeature={axis={wght=500,wdth=106}}},
      {Size={11-21.59},  RawFeature={axis={wght=380,wdth=100}}},
      {Size={21.6-},        RawFeature={axis={wght=300,wdth=95}}}
    },
  },
  ]

\begin{document}

\noindent {\tiny The quick brown fox jumps over the lazy dog.\\
  0123456789}

\noindent {\scriptsize The quick brown fox jumps over the lazy dog.}

\noindent {\footnotesize The quick brown fox jumps over the lazy dog.}

\noindent {\small The quick brown fox jumps over the lazy dog.}

\noindent {\normalsize The quick brown fox jumps over the lazy
  dog. The quick brown fox jumps over the lazy dog.
  The quick brown fox jumps over the lazy dog.\\
  0123456789\\
  ABCDEFGHIJKLMNOPQRSTUVWXYZ}

\noindent {\large The quick brown fox jumps over the lazy dog.}

\noindent {\Large The quick brown fox jumps over the lazy dog.}

\noindent {\LARGE The quick brown fox jumps over the lazy dog.}

\noindent {\huge The quick brown fox jumps over the lazy}

\end{document}
```
And got this strange effect, with the tops of letters cut off, but only in the first line of a paragraph:
<img width="230" alt="image" src="https://user-images.githubusercontent.com/16215767/148543668-76370707-a191-4fee-8961-386de30e1f6c.png">
The font's vertical metrics needed correcting, and I did that, but it didn't correct the problem. I'll do some research, but at the moment I'm at a loss.

The good news is that there seems to be no distortion of outlines. I'll probably make cff2-flavored variable fonts a standard part of the package.

**Reply by psb1558** [reply]
Works fine in FontGoggles and the Adobe suite. Okay in variable font test pages like [axis-praxis](https://www.axis-praxis.org/specimens/__DEFAULT__), but only in Chrome, not in Firefox (which I suppose doesn't recognize cff2-flavored variable fonts yet).

Letters not cut off when I switch to a ttf version of the font generated by fontmake, but the outline distortions are back.

**Comment by psb1558** [comment]
Changing the font's units per em from 2048 to 1000 fixes the problem:
<img width="262" alt="image" src="https://user-images.githubusercontent.com/16215767/148576358-3ba6e526-61b0-4ce8-9b5b-e6f724a9e1ad.png">
2048 is common for TrueType fonts. 2048 is **legal** for .otf fonts, but 1000 is most common, and some applications have not been tested with a higher UPM.

**Comment by psb1558** [comment]
Added the italic (reminder: very rough and incomplete!) and tested with these settings (suggestions/adjustments invited):
```
\setmainfont{JunicodeTwoBetaVF-Regular.otf}[
  ItalicFont = JunicodeTwoBetaVF-Italic.otf,
  BoldFont = JunicodeTwoBetaVF-Regular.otf,
  BoldItalicFont =  JunicodeTwoBetaVF-Italic.otf,
  Numbers={Lowercase,Proportional},
  UprightFeatures={
    SizeFeatures={
      {Size={-8.5},      RawFeature={axis={wght=550,wdth=115}}},
      {Size={8.6-10.99}, RawFeature={axis={wght=475,wdth=110}}},
      {Size={11-21.59},  RawFeature={axis={wght=400,wdth=100}}},
      {Size={21.6-},     RawFeature={axis={wght=300,wdth=95}}}
    },
  },
  ItalicFeatures={
    SizeFeatures={
      {Size={-8.5},      RawFeature={axis={wght=550,wdth=112}}},
      {Size={8.6-10.99}, RawFeature={axis={wght=475,wdth=106}}},
      {Size={11-21.59},  RawFeature={axis={wght=400,wdth=100}}},
      {Size={21.6-},     RawFeature={axis={wght=300,wdth=95}}}
    },
  },
  BoldFeatures={
    SizeFeatures={
      {Size={-8.5},      RawFeature={axis={wght=700,wdth=118}}},
      {Size={8.6-10.99}, RawFeature={axis={wght=700,wdth=112}}},
      {Size={11-21.59},  RawFeature={axis={wght=650,wdth=100}}},
      {Size={21.6-},     RawFeature={axis={wght=600,wdth=95}}}
    },
  },
  BoldItalicFeatures={
    SizeFeatures={
      {Size={-8.5},      RawFeature={axis={wght=700,wdth=112}}},
      {Size={8.6-10.99}, RawFeature={axis={wght=700,wdth=106}}},
      {Size={11-21.59},  RawFeature={axis={wght=650,wdth=100}}},
      {Size={21.6-},     RawFeature={axis={wght=600,wdth=95}}}
    },
  },
]
```

**Reply by adunning** [reply]
That is brilliant; thank you! I will very happily start testing that with a current project.

**Comment by psb1558** [comment]
Rendering of otf variable fonts in PDF viewers is still a little buggy, but perfectly legible, and they seem to print well. You've probably gotten the notice from the luaotfload issue thread that the bugs in ttf rendering have been fixed. I suppose the fixed version will be in the next TeXLive release (this spring?), and then the ttf version of the variable font will probably be the one to go with.

Meanwhile I hope you'll post any improvements here.

**Reply by adunning** [reply]
It does indeed look as if TeX Live 2021 has fixed the bugs!

**Comment by eclecticfluff** [comment]
I've been messing around with this for my use, but came across an issue, and a possible improvement. Firstly, for whatever reason fontspec throws unit errors when I specify base font size in the document class `\documentclass[12pt]{report}` Ultimately I got rid of the automatic adjustments so it would compile properly but it would be nice to have a proper solution. As for improvements, the enlarge axis could be abused to increase the x-height at small sizes. Anything above 10 or 15 starts looking wonky, but within that threshold I think it provides a nice bit of microtypography.
After looking at the output for a bit, it seems like the enlarge axis does not currently extend to ligatures, which is understandable considering its use case. it's subtle enough at such small enlarge values as is, but it's not too much of a chore, the output would look that much more pleasant with them properly supported.

**Reply by psb1558** [reply]
I've used Junicode with fontspec quite a bit and haven't seen a bug like this. Can you post the following:

a minimal file demonstrating the problem;
whether you're using XeLaTeX or LuaLaTex and the version;
version of Junicode used and where obtained.

Thanks.

Since the Enlarge axis is for sentence-initial characters likely to be used in medieval texts, i.e. isolated letters, what would be gained by implementing it for ligatures? (At present it is not implemented for ligatures, phonetic symbols, and in general anything that's not likely to appear in sentence-initial position in a medieval text. For letters that could appear in that environment, the collection is growing slowly.)

**Reply by eclecticfluff** [reply]
Junicode: Version 1.049, otf downloaded from this very github.
LuaTeX: version 1.15.0
![image](https://user-images.githubusercontent.com/65306291/166249096-3dab2cb2-a2ba-4437-bd97-4d3676a60695.png)
```
\documentclass{article}

\usepackage{fontspec}

\setmainfont{JunicodeTwoBetaVF-Roman.otf}[
UprightFeatures={
	RawFeature={axis={wght=400,wdth=100,enlarge=10}}
},
]
\begin{document}
	Quantify fifteen flighty friends.\\
	fff ftt fyy frr fll fii
\end{document}
```
I know that the enlarged minuscules are intended for encoding actual information, not aesthetic tweaks, hence saying that I was abusing the feature. I was thinking of using it as a general x-height axis, which would allow tweaks to the font for general aesthetics and enlarging the eyes of smaller point sizes for optical sizing. It's very much a nice to have, but not necessary thing for my decidedly non medievalist use case.

**Reply by eclecticfluff** [reply]
Wait--I just realized I posted the wrong code. get separated out into distinct posts. I'm trying to figure out a proper MWE for the error, but stuff is starting to look inconsistent between compiles which isn't great.

**Reply by eclecticfluff** [reply]
Tracked down the true culpret... 
```latex
\setmainfont{JunicodeTwoBetaVF-Roman.otf}[
UprightFeatures={
	SizeFeatures={
		{Size={-8.5},      RawFeature={axis={wght=550,wdth=115}}},
		{Size={8.6-10.99}, RawFeature={axis={wght=475,wdth=110}}},
		{Size={11--21.59},  RawFeature={axis={wght=400,wdth=100}}},
		{Size={21.6-},     RawFeature={axis={wght=300,wdth=95}}}
	},
},
]
```
compiles, only when a point size is not specified between 11 and 21.6 (note the extra hyphen). Thanks to error messages pointing me in a completely different direction and syntax highlighting that didn't notice anything amiss I didn't realize anything could be off.

**Reply by psb1558** [reply]
Okay, then, with this code it seems to be working fine (I added some more size tests):
```
\documentclass{article}

\usepackage{fontspec}

\setmainfont{JunicodeTwoBetaVF-Roman.ttf}[
UprightFeatures={
	SizeFeatures={
		{Size={-8.5},      RawFeature={axis={wght=550,wdth=115}}},
		{Size={8.6-10.99}, RawFeature={axis={wght=475,wdth=110}}},
		{Size={11-21.59},  RawFeature={axis={wght=400,wdth=100}}},
		{Size={21.6-},     RawFeature={axis={wght=300,wdth=95}}}
	},
},
]
\begin{document}
	Normal: Quantify fifteen flighty friends.

        \tiny Tiny: Quantify fifteen flighty friends.

        \small Small: Quantify fifteen flighty friends.

        \large Large: Quantify fifteen flighty friends.

        \huge huge: Quantify fifteen flighty friends.
        
\end{document}
```
<img width="461" alt="image" src="https://user-images.githubusercontent.com/16215767/166389809-b1267392-4112-43f5-a243-d7276c0d1cd8.png">

At present, the ttf variable font is yielding better results than the otf. Ttf:
<img width="151" alt="image" src="https://user-images.githubusercontent.com/16215767/166386676-8656604a-6c23-4785-8824-3d3fbb9fe353.png">
And otf:
<img width="147" alt="image" src="https://user-images.githubusercontent.com/16215767/166388349-1d7b1fc1-326f-4353-92dc-44b26a296b11.png">

**Comment by psb1558** [comment]
Okay, thanks for this info, @eclecticfluff. LuaTeX isn't *quite* ready for variable fonts yet. They've fixed the TTF outline problem in luaotfload, but you now get errors when certain positions along the axes are requested, and you've run afoul of that bug.

There is a fix in the [luaotfload dev branch](https://github.com/latex3/luaotfload/tree/dev), but I don't know when this will be available via an update to the current TeX distros.

What you're looking for is not the Enlarge axis but a true Optical Size axis. I'd supply one for Junicode if life were infinitely long, but alas, the job is too big (it would involve supplying at least one new master for each of more than 1000 glyphs).

My [Elstob font](https://github.com/psb1558/Elstob-font) has an Optical Size axis, and it is drawn from the same broad Fell family as Junicode (if you like that style). The TTF version of the variable font should work under TeXLive 2022, though it will throw some errors, as mentioned. The development version (I'm probably a couple of weeks away from a beta release) will also have a slant axis for the italic, since the slant of the original italic is rather extreme:
![Elstob-movie](https://user-images.githubusercontent.com/16215767/166290584-91eb9ac7-77e0-4e00-87de-8467bd16ba81.gif)

**Reply by eclecticfluff** [reply]
Wow! How many fonts are you actively working on? Elstob suits my needs perfectly, but I will miss the obscure fr and fty ligatures of Junicode as well as its subtle rounding of corners.

**Comment by psb1558** [comment]
Just Junicode and Elstob right now. Yeah, Elstob is meant to look more angular, and I keep nudging it more in that direction. Perhaps that gives it a more modern look than the extremely retro Junicode. If you use Elstob, perhaps you'll find design features you like that will make up for things you miss from Junicode (my personal favorites are the sloping cross-strokes and top serifs in the italic face—they come straight from the 17th-c. typeface).
<img width="250" alt="image" src="https://user-images.githubusercontent.com/16215767/166325531-5231ba03-50b9-406c-8985-2dc47f1799d7.png">
